Feature/Team

susilnem opened the pull request:
Addresses
- Add new team feature on project

## Changes
- Add new model for team
- Add new Archivable model
- set team on Project model  


## This PR doesn't introduce any:

- [ ] temporary files, auto-generated files or secret keys
- [ ] n+1 queries
- [ ] flake8 issues
- [ ] `print`
- [ ] typos
- [ ] unwanted comments

## This PR contains valid:

- [ ] tests
- [ ] permission checks (tests here too)
- [ ] translations

susilnem committed:
feat(team): Add new team feature to a project

tnagorra commented on apps/common/models.py:
```diff
@@ -44,3 +44,22 @@ class Meta(TypedModelMeta):  # type: ignore[reportIncompatibleVariableOverride]
+
+    class Meta(TypedModelMeta):  # type: ignore[reportIncompatibleVariableOverride]
+        abstract = True
+        ordering = ["-id"]
```
@thenav56 Do we need to add ordering here?

thenav56 replied:
Let's do that, to have a default ordering.

tnagorra commented on apps/contributor/admin.py:
```diff
@@ -1,9 +1,24 @@
+
+@admin.register(ContributorUser)
+class ContributorUserAdmin(DjangoQLSearchMixin, admin.ModelAdmin):
+    list_display = (
+        "user_id",
+        "username",
+    )
+    list_filter = ("team",)
```
can we also add the age of the account?

tnagorra commented on apps/contributor/admin.py:
```diff
@@ -1,9 +1,24 @@
+
+@admin.register(ContributorTeam)
+class ContributorTeamAdmin(DjangoQLSearchMixin, admin.ModelAdmin):
+    list_display = ("name", "is_archived")
```
Can we add age of the team?

tnagorra commented on apps/project/admin.py:
```diff
@@ -1,17 +1,48 @@
 
 
+class HasTeamFilter(admin.SimpleListFilter):
+    title = _("Has Team")
+    parameter_name = "has_team"
```
Maybe rename to "included in team"? @thenav56

thenav56 replied:
Sounds good

tnagorra commented on apps/common/models.py:
```diff
@@ -44,3 +44,22 @@ class Meta(TypedModelMeta):  # type: ignore[reportIncompatibleVariableOverride]
         return str(self.pk)
+
+
+class ArchivableResource(Model):
```
We also need to create a re-usable serializer for ArchivableResource. Similar to UserResource

We can then replace the logic for populating archived_at and archived_by for all the resources that are archivable.
The logic is already implemented for ContributorUserGroup

susilnem committed:
feat(team): Add query for team

susilnem committed:
feat(common): Add archivable resources generic serializer

- Add mixing for ArchivableResource
- Udpate on ContributorUserGroup
- Add test cases for archive on ContributorUserGroup

tnagorra commented on apps/common/serializers.py:
```diff
@@ -58,3 +59,36 @@ def update(self, instance: ModelType, validated_data: dict[str, typing.Any]) ->
+):
+    instance: ModelType | None  # type: ignore[override]
+
+    @property
+    def context(self) -> ContextType:  # type: ignore[override]
+        ctx = super().context
+        assert ctx is not None, f"Always pass context when using {type(self)}"
+        return typing.cast("ContextType", ctx)
+
+    def _set_archive_fields(self, validated_data: dict[str, typing.Any], instance: ModelType) -> None:
```
We need to add "is_archived" field here so that it does not need to be repeated in each class where it's used

tnagorra committed:
fix(serializer): fix typing issues for project and tutorial status
